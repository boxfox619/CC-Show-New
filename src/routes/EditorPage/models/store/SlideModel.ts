import AssetModel from 'src/core/models/AssetModel';

export default class SlideModel {
    constructor(
        public id: number = 0,
        public name: string = '',
        public thumbnail?: string,
        public assets: AssetModel[] = [],
        public lastAssetId: number = 0,
        public selectedAssetId?: number
    ){}
}