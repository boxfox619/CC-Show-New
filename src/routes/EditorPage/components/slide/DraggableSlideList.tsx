import * as React from 'react';
import SlideModel from './../../models/store/SlideModel';
import SlidePreview from './SlidePreview';
import SlideCreateCard from './SlideCreateCard';

interface Props {
    slides: SlideModel[],
    selectedSlideId: number,
    moveSlide: (payload: {from: number, to: number}) => void,
    shareSlide: (id: number) => void,
    copySlide: (id: number) => void,
    deleteSlide: (id: number) => void,
    selectSlide: (id: number) => void,
    createSlide: () => void
}

const placeholder = document.createElement("li");
placeholder.className = "placeholder";

export default class DraggableSlideList extends React.Component<Props> {
    dragged?: any = undefined;
    over?: any = undefined;
    nodePlacement?: any = undefined;

    public render() {
        const renderSlidePreviews = (slides: SlideModel[]) => {
            return slides.map((slide, idx) => {
                const isActive = idx === this.props.selectedSlideId;
                const onShare = () => this.props.shareSlide(slide.id);
                const onCopy = () => this.props.copySlide(slide.id);
                const onDelete = () => this.props.deleteSlide(slide.id);
                const onClick = () => this.props.selectSlide(slide.id);
                return (
                    <li data-id={idx}
                        key={'slide' + slide.id}
                        draggable={true}
                        onDragEnd={this.dragEnd}
                        onDragStart={this.dragStart}>
                        <SlidePreview
                            idx={idx}
                            slide={slide}
                            active={isActive}
                            onShare={onShare}
                            onCopy={onCopy}
                            onDelete={onDelete}
                            onClick={onClick}
                        />
                    </li>
                )
            })
        };
        return (
            <>
                <ul onDragOver={this.dragOver}>
                    {renderSlidePreviews(this.props.slides)}
                </ul>
                <SlideCreateCard onClick={this.props.createSlide}/>
            </>
        )
    }

    dragOver = (e: React.DragEvent<HTMLElement>) => {
        e.preventDefault();
        if (!this.dragged) {
            return;
        }
        this.dragged.style.display = "none";
        let target = e.target as HTMLElement;
        if (target.className === "placeholder" || target.tagName === 'UL') {
            return;
        }
        while (target.tagName !== 'LI') {
            target = target.parentNode as HTMLElement;
        }
        this.over = target;
        const relY = e.clientY - this.over.offsetTop;
        const height = this.over.offsetHeight / 2;
        let parent = e.target as HTMLElement;
        while (parent.tagName !== 'UL') {
            parent = parent.parentNode as HTMLElement;
        }
        while (target.tagName !== 'LI') {
            target = target.parentNode as HTMLElement;
        }
        if (relY > height) {
            this.nodePlacement = "after";
            parent.insertBefore(placeholder, target.nextElementSibling);
        }
        else if (relY < height) {
            this.nodePlacement = "before"
            parent.insertBefore(placeholder, target);
        }
    }
    dragEnd = (e: React.DragEvent<HTMLElement>) => {
        if (!this.dragged || !this.over) {
            return;
        }
        this.dragged.style.display = "block";
        if (!this.dragged.parentNode.contains(placeholder)) {
            return;
        }
        this.dragged.parentNode.removeChild(placeholder);
        const from = Number(this.dragged.dataset.id);
        let to = Number(this.over.dataset.id);
        if (from < to) {
            to--
        };
        if (this.nodePlacement === "after") {
            to++
        };
        this.props.moveSlide({from, to});
    }

    dragStart = (e: React.DragEvent<HTMLElement>) => {
        let target = e.currentTarget as HTMLElement;
        while (target.tagName !== 'LI') {
            target = target.parentNode as HTMLElement;
        }
        this.dragged = target;
        e.dataTransfer.effectAllowed = 'move';
        e.dataTransfer.setData("text/html", e.currentTarget.innerHTML);
    }
}