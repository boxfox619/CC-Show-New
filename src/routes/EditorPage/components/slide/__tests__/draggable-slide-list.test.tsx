import * as React from 'react';
import * as Enzyme from 'enzyme';
import DraggableSlideList from '../DraggableSlideList';
import SlideModel from '../../../models/store/SlideModel';

describe('<IconButton />', () => {
  const url = 'https://avatars0.githubusercontent.com/u/14067209?s=88&v=4';
  const mockFn = jest.fn();
  it('should match snapshot', () => {
    const wrapper = Enzyme.shallow(
      <DraggableSlideList
        selectedSlideId={0}
        slides={[new SlideModel(0, '테스트 슬라이드', url), new SlideModel(1, '테스트 슬라이드2', url)]}
        exchangeSlide={mockFn} 
        selectSlide={mockFn}
        copySlide={mockFn}
        createSlide={mockFn}
        shareSlide={mockFn}
        deleteSlide={mockFn}
        />
      );
    expect(wrapper).toMatchSnapshot();
  });
});