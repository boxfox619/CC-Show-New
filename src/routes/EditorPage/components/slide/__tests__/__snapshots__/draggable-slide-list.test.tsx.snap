// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DraggableSlideList /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DraggableSlideList
    copySlide={[MockFunction]}
    createSlide={[MockFunction]}
    deleteSlide={[MockFunction]}
    moveSlide={[MockFunction]}
    selectSlide={[MockFunction]}
    selectedSlideId={0}
    shareSlide={[MockFunction]}
    slides={
      Array [
        Slide {
          "assets": Array [],
          "id": 0,
          "lastAssetId": 0,
          "name": "테스트 슬라이드",
          "selectedAssetId": undefined,
          "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
        },
        Slide {
          "assets": Array [],
          "id": 1,
          "lastAssetId": 0,
          "name": "테스트 슬라이드2",
          "selectedAssetId": undefined,
          "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ul
          onDragOver={[Function]}
        >
          <li
            data-id={0}
            draggable={true}
            onDragEnd={[Function]}
            onDragStart={[Function]}
          >
            <SlidePreview
              active={true}
              idx={0}
              onClick={[MockFunction]}
              onCopy={[MockFunction]}
              onDelete={[MockFunction]}
              onShare={[MockFunction]}
              slide={
                Slide {
                  "assets": Array [],
                  "id": 0,
                  "lastAssetId": 0,
                  "name": "테스트 슬라이드",
                  "selectedAssetId": undefined,
                  "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                }
              }
            />
          </li>
          <li
            data-id={1}
            draggable={true}
            onDragEnd={[Function]}
            onDragStart={[Function]}
          >
            <SlidePreview
              active={false}
              idx={1}
              onClick={[MockFunction]}
              onCopy={[MockFunction]}
              onDelete={[MockFunction]}
              onShare={[MockFunction]}
              slide={
                Slide {
                  "assets": Array [],
                  "id": 1,
                  "lastAssetId": 0,
                  "name": "테스트 슬라이드2",
                  "selectedAssetId": undefined,
                  "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                }
              }
            />
          </li>
        </ul>,
        <SlideCreateCard
          onClick={[MockFunction]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <li
              data-id={0}
              draggable={true}
              onDragEnd={[Function]}
              onDragStart={[Function]}
            >
              <SlidePreview
                active={true}
                idx={0}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 0,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />
            </li>,
            <li
              data-id={1}
              draggable={true}
              onDragEnd={[Function]}
              onDragStart={[Function]}
            >
              <SlidePreview
                active={false}
                idx={1}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 1,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드2",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />
            </li>,
          ],
          "onDragOver": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "slide0",
            "nodeType": "host",
            "props": Object {
              "children": <SlidePreview
                active={true}
                idx={0}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 0,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />,
              "data-id": 0,
              "draggable": true,
              "onDragEnd": [Function],
              "onDragStart": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": true,
                "idx": 0,
                "onClick": [MockFunction],
                "onCopy": [MockFunction],
                "onDelete": [MockFunction],
                "onShare": [MockFunction],
                "slide": Slide {
                  "assets": Array [],
                  "id": 0,
                  "lastAssetId": 0,
                  "name": "테스트 슬라이드",
                  "selectedAssetId": undefined,
                  "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "slide1",
            "nodeType": "host",
            "props": Object {
              "children": <SlidePreview
                active={false}
                idx={1}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 1,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드2",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />,
              "data-id": 1,
              "draggable": true,
              "onDragEnd": [Function],
              "onDragStart": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": false,
                "idx": 1,
                "onClick": [MockFunction],
                "onCopy": [MockFunction],
                "onDelete": [MockFunction],
                "onShare": [MockFunction],
                "slide": Slide {
                  "assets": Array [],
                  "id": 1,
                  "lastAssetId": 0,
                  "name": "테스트 슬라이드2",
                  "selectedAssetId": undefined,
                  "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "li",
          },
        ],
        "type": "ul",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ul
            onDragOver={[Function]}
          >
            <li
              data-id={0}
              draggable={true}
              onDragEnd={[Function]}
              onDragStart={[Function]}
            >
              <SlidePreview
                active={true}
                idx={0}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 0,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />
            </li>
            <li
              data-id={1}
              draggable={true}
              onDragEnd={[Function]}
              onDragStart={[Function]}
            >
              <SlidePreview
                active={false}
                idx={1}
                onClick={[MockFunction]}
                onCopy={[MockFunction]}
                onDelete={[MockFunction]}
                onShare={[MockFunction]}
                slide={
                  Slide {
                    "assets": Array [],
                    "id": 1,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드2",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  }
                }
              />
            </li>
          </ul>,
          <SlideCreateCard
            onClick={[MockFunction]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li
                data-id={0}
                draggable={true}
                onDragEnd={[Function]}
                onDragStart={[Function]}
              >
                <SlidePreview
                  active={true}
                  idx={0}
                  onClick={[MockFunction]}
                  onCopy={[MockFunction]}
                  onDelete={[MockFunction]}
                  onShare={[MockFunction]}
                  slide={
                    Slide {
                      "assets": Array [],
                      "id": 0,
                      "lastAssetId": 0,
                      "name": "테스트 슬라이드",
                      "selectedAssetId": undefined,
                      "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                    }
                  }
                />
              </li>,
              <li
                data-id={1}
                draggable={true}
                onDragEnd={[Function]}
                onDragStart={[Function]}
              >
                <SlidePreview
                  active={false}
                  idx={1}
                  onClick={[MockFunction]}
                  onCopy={[MockFunction]}
                  onDelete={[MockFunction]}
                  onShare={[MockFunction]}
                  slide={
                    Slide {
                      "assets": Array [],
                      "id": 1,
                      "lastAssetId": 0,
                      "name": "테스트 슬라이드2",
                      "selectedAssetId": undefined,
                      "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                    }
                  }
                />
              </li>,
            ],
            "onDragOver": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "slide0",
              "nodeType": "host",
              "props": Object {
                "children": <SlidePreview
                  active={true}
                  idx={0}
                  onClick={[MockFunction]}
                  onCopy={[MockFunction]}
                  onDelete={[MockFunction]}
                  onShare={[MockFunction]}
                  slide={
                    Slide {
                      "assets": Array [],
                      "id": 0,
                      "lastAssetId": 0,
                      "name": "테스트 슬라이드",
                      "selectedAssetId": undefined,
                      "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                    }
                  }
                />,
                "data-id": 0,
                "draggable": true,
                "onDragEnd": [Function],
                "onDragStart": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "active": true,
                  "idx": 0,
                  "onClick": [MockFunction],
                  "onCopy": [MockFunction],
                  "onDelete": [MockFunction],
                  "onShare": [MockFunction],
                  "slide": Slide {
                    "assets": Array [],
                    "id": 0,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "slide1",
              "nodeType": "host",
              "props": Object {
                "children": <SlidePreview
                  active={false}
                  idx={1}
                  onClick={[MockFunction]}
                  onCopy={[MockFunction]}
                  onDelete={[MockFunction]}
                  onShare={[MockFunction]}
                  slide={
                    Slide {
                      "assets": Array [],
                      "id": 1,
                      "lastAssetId": 0,
                      "name": "테스트 슬라이드2",
                      "selectedAssetId": undefined,
                      "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                    }
                  }
                />,
                "data-id": 1,
                "draggable": true,
                "onDragEnd": [Function],
                "onDragStart": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "active": false,
                  "idx": 1,
                  "onClick": [MockFunction],
                  "onCopy": [MockFunction],
                  "onDelete": [MockFunction],
                  "onShare": [MockFunction],
                  "slide": Slide {
                    "assets": Array [],
                    "id": 1,
                    "lastAssetId": 0,
                    "name": "테스트 슬라이드2",
                    "selectedAssetId": undefined,
                    "thumbnail": "https://avatars0.githubusercontent.com/u/14067209?s=88&v=4",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "li",
            },
          ],
          "type": "ul",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
